File: fitbitsdk/src/main/java/com/mindbodyonline/fitbitsdk/authentication/ui/FitbitLoginActivity.java
package com.mindbodyonline.fitbitsdk.authentication.ui;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.webkit.WebView;

import com.mindbodyonline.fitbitsdk.R;
import com.mindbodyonline.fitbitsdk.authentication.AuthenticationHandler;
import com.mindbodyonline.fitbitsdk.authentication.AuthenticationResult;
import com.mindbodyonline.fitbitsdk.authentication.AuthorizationController;
import com.mindbodyonline.fitbitsdk.authentication.ClientCredentials;
import com.mindbodyonline.fitbitsdk.authentication.Scope;

import java.util.HashSet;
import java.util.Set;


public class FitbitLoginActivity extends AppCompatActivity implements AuthenticationHandler {

    public static final String CONFIGURATION_VERSION = "CONFIGURATION_VERSION";
    public static final String AUTHENTICATION_RESULT_KEY = "AUTHENTICATION_RESULT_KEY";
    private static final String CLIENT_CREDENTIALS_KEY = "CLIENT_CREDENTIALS_KEY";
    private static final String SCOPES_KEY = "SCOPES_KEY";
    private WebView loginWebview;

    public static Intent createIntent(Context context, @NonNull ClientCredentials clientCredentials, Set<Scope> scopes) {
        return createIntent(context, null, clientCredentials, scopes);
    }

    public static Intent createIntent(Context context, Integer configVersion, @NonNull ClientCredentials clientCredentials, Set<Scope> scopes) {
        Intent intent = new Intent(context, FitbitLoginActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
        intent.putExtra(CLIENT_CREDENTIALS_KEY, clientCredentials);
        intent.putExtra(CONFIGURATION_VERSION, configVersion);
        intent.putExtra(SCOPES_KEY, scopes.toArray(new Scope[scopes.size()]));

        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        loginWebview = findViewById(R.id.login_webview);

        ClientCredentials clientCredentials = getIntent().getParcelableExtra(CLIENT_CREDENTIALS_KEY);
        Parcelable[] scopes = getIntent().getParcelableArrayExtra(SCOPES_KEY);
        Set<Scope> scopesSet = new HashSet<>();
        for (Parcelable scope : scopes) {
            scopesSet.add((Scope) scope);
        }

        AuthorizationController authorizationController = new AuthorizationController(
                loginWebview,
                clientCredentials,
                this);

        authorizationController.authenticate(scopesSet);

    }

    @Override
    public void onAuthFinished(AuthenticationResult result) {
        loginWebview.setVisibility(View.GONE);

        Intent resultIntent = new Intent();
        resultIntent.putExtra(AUTHENTICATION_RESULT_KEY, result);
        resultIntent.putExtra(CONFIGURATION_VERSION, getIntent().getIntExtra(CONFIGURATION_VERSION, 0));
        setResult(Activity.RESULT_OK, resultIntent);
        finish();
    }
}

File: app/src/main/java/com/mindbodyonline/fitbitintegration/FitbitLoginActivity.kt
package com.mindbodyonline.fitbitintegration

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.util.Log
import android.view.View
import com.mindbodyonline.fitbitsdk.authentication.AuthenticationHandler
import com.mindbodyonline.fitbitsdk.authentication.AuthenticationManager
import com.mindbodyonline.fitbitsdk.authentication.AuthenticationResult
import com.mindbodyonline.fitbitsdk.service.FitbitService
import com.mindbodyonline.fitbitsdk.service.models.UserProfile
import kotlinx.android.synthetic.main.activity_fitbit_login.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class FitbitLoginActivity : AppCompatActivity(), AuthenticationHandler {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_fitbit_login)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (!AuthenticationManager.onActivityResult(requestCode, resultCode, data, this)) {
            // Handle other activity results, if needed
        }
    }
    override fun onResume() {
        super.onResume()
        if (AuthenticationManager.isLoggedIn()) {
            onLoggedIn()
            button.visibility = View.GONE
        }else{
            button.visibility = View.VISIBLE
            button.setOnClickListener { _ -> AuthenticationManager.login(this) }
        }
    }

    override fun onAuthFinished(authenticationResult: AuthenticationResult?) {
        if (authenticationResult != null) {
            if (authenticationResult.isSuccessful) {
                onLoggedIn()
            } else {
                displayAuthError(authenticationResult)
            }
        }
    }

    fun onLoggedIn(){
        val sharedPreferences = getSharedPreferences("Fitbit", Context.MODE_PRIVATE)
        val fService = FitbitService(sharedPreferences, AuthenticationManager.getAuthenticationConfiguration().clientCredentials)
        fService.getUserService().profile().enqueue(object : Callback<UserProfile> {
            override fun onResponse(call: Call<UserProfile>?, response: Response<UserProfile>?) {
                Log.d("Service", "Response: " + response?.body()?.user?.dateOfBirth)
            }

            override fun onFailure(call: Call<UserProfile>?, t: Throwable?) {
                Log.d("Service", "Error: " + t?.message)
            }

        })
    }

    private fun displayAuthError(authenticationResult: AuthenticationResult) {
        var message = ""

        when (authenticationResult.status) {
            AuthenticationResult.Status.dismissed -> message = getString(R.string.login_dismissed)
            AuthenticationResult.Status.error -> message = authenticationResult.errorMessage
            AuthenticationResult.Status.missing_required_scopes -> {
                val missingScopes = authenticationResult.missingScopes
                val missingScopesText = TextUtils.join(", ", missingScopes)
                message = getString(R.string.missing_scopes_error) + missingScopesText
            }
        }

        AlertDialog.Builder(this)
                .setTitle(R.string.login_title)
                .setMessage(message)
                .setCancelable(false)
                .setPositiveButton(android.R.string.yes, DialogInterface.OnClickListener { dialog, id -> })
                .create()
                .show()
    }

}
